web/src/OverviewLogPane.tsx

```diff
@@ -320,27 +372,150 @@ export class OverviewLogComponent extends Component<OverviewLogComponentProps> {
       }
       return true
     })
+
     this.logCheckpoint = patch.checkpoint
+    lines.forEach((line) => this.lineHashList.append(line))
+
+    if (startCheckpoint) {
+      // If this is an incremental render, put the lines in the forward buffer.
+      lines.forEach((line) => {
+        this.forwardBuffer.push(line)
+      })
+    } else {
+      // If this is the first render, put the lines in the backward buffer, so
+      // that the last lines get rendered first.
+      lines.forEach((line) => {
+        this.backwardBuffer.push(line)
+      })
+    }
+
+    if (this.renderBufferRafId) return
+    this.renderBufferRafId = this.props.raf.requestAnimationFrame(
+      this.renderBuffer
+    )
+  }
+
+  // We have two render buffers:
+  // - a buffer of newer logs that we haven't rendered yet.
+  // - a buffer of older logs that we haven't rendered yet.
+  // First, process the newer logs.
+  // If we're out of new logs to render, go back through the old logs.
```

Comment:
understanding check: a single call of this function renders at most n (where n = renderWindow) lines of forward buffer and n lines of backward buffer, and then calls itself recursively if there's more stuff to render?

Reply:
yep! it's yielding the CPU and then scheduling another render job